File: graphs_greedy/graph_detect_cycle.cpp
#include <iostream>
#include <list>
#include <vector>
#include <unordered_map>

using namespace std;

class Edge {
    private:
        int src;
        int dest;
        int weight;
    public:
        Edge(int src, int dest, int weight) {
            this->src = src;
            this->dest = dest;
            this->weight = weight;
        }
        int getSrc() {
            return this->src;
        }
        int getDest() {
            return this->dest;
        }
};

class Graph {
    private:
        int V;
        int E;
        list<int> *adj;
        list<int> *edge;

        bool directedGraphIsCyclicUtil(int v, bool visited[], int recStack[], vector<int> path) {


            if(visited[v] == false) {
                // Mark the current node as visited and part of recursion stack
                visited[v] = true;
                recStack[v] = v;
                path.push_back(v);

                // Recur for all the vertices adjacent to this vertex
                list<int>::iterator i;
                for(i = adj[v].begin(); i != adj[v].end(); ++i) {
                    if(!visited[*i] && directedGraphIsCyclicUtil(*i, visited, recStack, path)) {
                        return true;
                    } else if(recStack[*i] >= 0) {
                        path.push_back(*i);
                        cout << "Cycle at node: " << v << " " << recStack[*i] << endl;
                        for(int i = 0; i < path.size(); i++) {
                            cout << path[i] << " ";
                        }
                        cout << endl;
                        return true;
                    }
                }
            }

            // Backtrack, remove vertex from recursion stack, and not in cyclic path
            recStack[v] = -1;
            path.pop_back();
            return false;
        }

    public:
        Graph(int V, int E) {
            this->V = V;
            this->E = E;
            this->adj = new list<int>[V];
            this->edge = new list<int>[V];
        }

        void addDirectedEdge(int u, int w) {
            this->adj[u].push_back(w);
        }

        void addUndirectedEdge(int u, int w) {
            this->adj[u].push_back(w);
            this->adj[w].push_back(u);

            // check if edge exists
            list<int>::iterator it;
            for(it = edge[u].begin(); it != edge[u].end(); it++) {
                if(*it == w) {
                    return;
                }
            }
            for(it = edge[w].begin(); it != edge[w].end(); it++) {
                if(*it == u) {
                    return;
                }
            }

            // edge not found, add it
            this->edge[u].push_back(w);
            return;
        }

        void constructDirectedGraph() {

            //    ________
            //   /        \
            //  /          V
            // 0 --> 1 --> 2
            // ^          /
            //  \________/
            //
            // 3 ----
            // ^    |
            // |____|

            addDirectedEdge(0, 1);
            addDirectedEdge(0, 2);
            addDirectedEdge(1, 2);
            addDirectedEdge(2, 0);
            addDirectedEdge(3, 3);
        }

        void constructUndirectedGraph() {

            // 1
            // |
            // |
            // 0 --- 2 --- 4
            // |           |
            // |____ 3 ____|

            addUndirectedEdge(0, 1);
            addUndirectedEdge(0, 2);
            addUndirectedEdge(0, 3);
            addUndirectedEdge(3, 4);
            addUndirectedEdge(2, 4);
        }

        void directedGraphIsCyclic() {

            // Path of cycle
            vector<int> path;

            // Mark all vertices as not visited and not part of recursion stack
            bool visited[V];
            int recStack[V];
            for(int i = 0; i < V; i++) {
                visited[i] = false;
                recStack[i] = -1;
            }

            // Call recursive helper function to detect cycle in different DFS trees
            for(int i = 0; i < V; i++) {
                if(!visited[i]) {
                    if(!directedGraphIsCyclicUtil(i, visited, recStack, path)) {
                        cout << "Graph doesn't contain cycle" << endl;
                    }
                }
            }
        }

        void undirectedGraphIsCyclicUsingDFS() {

            vector<int> path;
            // Mark all vertices as not visited
            bool visited[V];
            for(int i = 0; i < V; i++) {
                visited[i] = false;
            }

            // in undirected graph, parent will always from a loop
            // We need to keep a node's parent updated to avoid check for cycle with parent
            int parent = -1;

            // Call recursive helper function to detect cycle in different DFS trees
            for(int i = 0; i < V; i++) {
                if(!visited[i]) {
                    if(undirectedGraphIsCyclicUtilUsingDFS(i, visited, path, parent)) {
                        cout << "Graph contains cycle" << endl;
                        cout << endl;
                    } else {
                        cout << "Graph doesn't contain cycle" << endl;
                    }
                }
            }
        }

        bool undirectedGraphIsCyclicUtilUsingDFS(int v, bool visited[], vector<int> path, int parent) {

            visited[v] = true;
            path.push_back(v);

            // Recur for all the vertices adjacent to this vertex
            list<int>::iterator i;
            for(i = adj[v].begin(); i != adj[v].end(); i++) {
                if(!visited[*i]) {
                    if(undirectedGraphIsCyclicUtilUsingDFS(*i, visited, path, v)) {
                        return true;
                    }
                } else if(*i != parent) {
                    path.push_back(*i);
                    for(int i = 0; i < path.size(); i++) {
                        cout << path[i] << " ";
                    }
                    cout << endl;
                    return true;
                }

            }
            path.pop_back();
            return false;
        }

        /**
         * Print cycle
         */
        void printPath(vector<Edge> path, int dest) {

            int path_size = path.size();
            bool visited[path_size];
            for(int i = 0; i < path_size; i++) {
                visited[i] = false;
            }

            int src = -1;
            // initialize first edge
            for(int i = 0; i < path_size; i++) {
                int curr_src = path[i].getSrc();
                int curr_dest = path[i].getDest();
                if(curr_src == dest || curr_dest == dest) {
                    visited[i] = true;
                    if(curr_src == dest) {
                        src = curr_dest;
                    } else {
                        src = curr_src;
                    }
                    break;
                }
            }

            vector<int> new_path;
            new_path.push_back(dest);
            new_path.push_back(src);

            bool not_found = true;
            int last_src = -1;
            while(src != dest) {

                for(int i = 0; i < path_size; i++) {
                    if(!visited[i]) {
                        if(path[i].getSrc() == src) {
                            visited[i] = true;
                            last_src = src;
                            src = path[i].getDest();
                            not_found = false;
                            break;
                        } else if(path[i].getDest() == src) {
                            visited[i] = true;
                            last_src = src;
                            src = path[i].getSrc();
                            not_found = false;
                            break;
                        }
                    }
                }

                if(not_found) {
                    src = last_src;
                    new_path.pop_back();
                } else {
                    new_path.push_back(src);
                    not_found = true;
                }
            }
            for(int i = 0; i < new_path.size(); i++) {
                cout << new_path[i] << " ";
            }
            cout << endl;
        }

        /**
         * Takes a vertex and finds its main parent
         */
        int find(int parent[], int i) {
            if(parent[i] == -1) {
                return i;
            }
            return find(parent, parent[i]);
        }

        /**
         * Takes parents and merges them one set
         */
        void unionVertices(int parent[], int src_parent, int dest_parent) {
            int src_set = find(parent, src_parent);
            int dest_set = find(parent, dest_parent);
            if(src_set >= 0) {
                parent[src_set] = dest_set;
            }
        }

        /**
         * Detect cycle in an undirected graph using disjoint set (or union-find)
         *
         * Uses a disjoint set data structure, this data structure keeps track
         * of set of elements partioned into a number of disjoint subsets
         *
         * This algo performs 2 functions:
         *   - Find:  Find which subset a particular element is in
         *            Used to determine if 2 elements are in the same subset
         *   - Union: Join 2 subsets into a single subset
         *
         * How it works:
         *   - Keep track of the subsets in an array called parent[]
         *      - Initially, all slots of parent array are initialized to -1
         *      - This means there is only one item in every subset
         *
         *   - Now process all edges one by one
         *      - For each edge, make subsets using both the vertices of the edge
         *      - If both the vertices are in the same subset, a cycle is found
         *      - If vertices are in different subset, take union of them. Make node 0 parent of node 1
         */
        void undirectedGraphIsCyclicUsingUnionFind() {

            int parent[V];
            for(int i = 0; i < V; i++) {
                parent[i] = -1;
            }

            vector<Edge> path;

            list<int>::iterator it;
            // Iterate through all edges of graph
            for(int i = 0; i < E; i++) {

                for(it = edge[i].begin(); it != edge[i].end(); it++) {

                    Edge temp(i, *it, 1);
                    path.push_back(temp);

                    int src = i;
                    int dest = *it;
                    int src_parent = find(parent, i);
                    int dest_parent = find(parent, *it);

                    // both points in same set
                    if(src_parent == dest_parent) {
                        //cout << src << " " << dest << " " << src_parent << " " << dest_parent << endl;
                        cout << "Found cycle - caused by edge " << src << "-" << dest << endl;
                        printPath(path, dest);
                        cout << endl;
                        break;
                    }

                    // if points in not the same set, then union them into same set
                    unionVertices(parent, src_parent, dest_parent);
                }
            }
        }
};

int main() {

    /*
     * Detect cycle in directed graph
     *
     * Algo: DFS, tarjan's algo, graph coloring
     * DFS: Time complexity: O(V + E), Space complexity O(V) -> O(V) path of cycle, O(V) recursion stack, O(V) visited array
     */
    Graph g_directed(4, 5);
    g_directed.constructDirectedGraph();
    g_directed.directedGraphIsCyclic();

    cout << "-------" << endl;

    /*
     * Detect cycle in undirected graph
     *
     * Algo: Union-find, DFS, BFS, and graph coloring
     */
    Graph g_undirected1(5, 5);
    g_undirected1.constructUndirectedGraph();

    /*
     * Union-find: Time complexity: O(E logV), Space complexity: O(V) -> O(V) path of cycle, O(V) recursion stack, O(V) visited array
     */
    g_undirected1.undirectedGraphIsCyclicUsingUnionFind();

    /*
     * DFS, BFS, Graph coloring: Time complexity: O(V + E), Space complexity: O(V) -> O(V) path of cycle, O(V) recursion stack, O(V) visited array
     */
    g_undirected1.undirectedGraphIsCyclicUsingDFS();

    return 0;
}

File: graphs_greedy/graph_detect_cycle.py
"""
Make a simple directed graph with cycle
Make a simple undirected graph with cycle

DFS on both to detect the cycle
"""

# -*- encoding: utf-8 *-*


class Node:
    """
    represents a node of the graph
    """
    def __init__(self, data):
        self.data = data


class Edge:
    """
    represent an edge of the graph
    """
    def __init__(self, src, dest, weight):
        self.src = src
        self.dest = dest
        self.weight = weight

    def getSrc(self):
        return self.src

    def getDest(self):
        return self.dest


class Graph:
    """
    represents a directed graph using adjacency list representation
              6 <-------- 5
              ^           ^
              |           |
              0 --> 1 --> 4 <---
                \  /      ^     |
                 vv       |     |
         3 <---- 2--------      |
         |                      |
          ----------------------
    """
    def __init__(self, total_vertices):
        self.total_vertices = total_vertices
        self.adj = {}

    def print_adj_list(self):
        print(self.adj)

    def construct_directed_graph(self):
        self.add_directed_edge(0, 1)
        self.add_directed_edge(0, 2)
        self.add_directed_edge(0, 6)
        self.add_directed_edge(1, 2)
        self.add_directed_edge(1, 4)
        self.add_directed_edge(2, 3)
        self.add_directed_edge(2, 4)
        self.add_directed_edge(3, 4)
        self.add_directed_edge(4, 5)
        self.add_directed_edge(5, 6)

    def construct_undirected_graph(self):
        """
                  6 ---------|
                  |          |
                  |          |
                  0 -- 1     |
                  |  /       |
                  | /        |
                  2 -- 4 --- 5
                  |  /
                  | /
                  3
        """
        self.add_directed_edge(0, 1)
        self.add_directed_edge(1, 0)
        self.add_directed_edge(0, 2)
        self.add_directed_edge(2, 0)
        self.add_directed_edge(1, 2)
        self.add_directed_edge(2, 1)
        self.add_directed_edge(1, 4)
        self.add_directed_edge(4, 1)
        self.add_directed_edge(0, 6)
        self.add_directed_edge(6, 0)
        self.add_directed_edge(2, 3)
        self.add_directed_edge(3, 2)
        self.add_directed_edge(3, 4)
        self.add_directed_edge(4, 3)
        self.add_directed_edge(4, 5)
        self.add_directed_edge(5, 4)
        self.add_directed_edge(5, 6)
        self.add_directed_edge(6, 5)
        self.add_directed_edge(2, 4)
        self.add_directed_edge(4, 2)

    def add_directed_edge(self, start, end):
        if start in self.adj:
            start_edge_list = self.adj[start]
            for index in range(len(start_edge_list)):
                if start_edge_list[index] == end:
                    return
            start_edge_list.append(end)
            self.adj[start] = start_edge_list
        else:
            self.adj[start] = [end]
        return

    def add_undirected_edge(self, start, end):
        self.add_directed_edge(start, end)
        self.add_directed_edge(end, start)

    def undirected_graph_detect_cycle_DFS_util(self, curr_node, visited, path, parent):
        visited[curr_node] = True
        path.append(curr_node)

        # Recur for all adjacent vertices
        if curr_node in self.adj:
            end_node_list = self.adj[curr_node]
            for adj_node in end_node_list:
                if not visited[adj_node]:
                    if self.undirected_graph_detect_cycle_DFS_util(adj_node, visited, path, curr_node):
                        return True
                elif adj_node != parent:
                    path.append(adj_node)
                    path.append(parent)
                    return True
        path.pop(-1)
        return False

    def undirected_graph_detect_cycle_DFS(self):
        path = []
        visited = [False] * self.total_vertices

        # In an undirected graph, parent will from a loop
        # We need to keep a node's parent updated to avoid check for cycle with parent
        parent = -1

        # Call recursive helper function to detect cycle in different DFS trees
        for curr_node in range(self.total_vertices):
            if not visited[curr_node]:
                if self.undirected_graph_detect_cycle_DFS_util(curr_node, visited, path, parent):
                    print("Found a cycle with DFS", path)
                    break
                else:
                    print("Graph doesn't contain cycle")
        return

    def find(self, parent, node):
        """
        Takes a vertex and finds its main parent
        """
        if parent[node] == -1:
            return node
        return self.find(parent, parent[node])

    def union_vertices(self, parent, src_parent, dest_parent):
        """
        Takes parents and merges them into one set
        """
        src_set = self.find(parent, src_parent)
        dest_set = self.find(parent, dest_parent)
        if src_set >= 0:
            parent[src_set] = dest_set

    def undirected_graph_detect_cycle_union_find(self):
        """
        Detect cycle in an undirected graph using disjoint set (or union-find)

        Uses a disjoint set data structure, this data structure keeps track of set of elements partitioned
        into a number of disjoint subsets

        This algo performs 2 functions:
        - Find:  Find which subset a particular element is in. Used to determine if 2 elements are in the same subset
        - Union: Join 2 subsets into a single subset

        How it works:
        - Keep track of the subsets in an array called parent[]
        - Initially, all slots of parent array are initialized to -1
        - This means there is only one item in every subset

        - Now process all edges one by one
            - For each edge, make subsets using both the vertices of the edge
            - If both the vertices are in the same subset, a cycle is found
            - If vertices are in different subset, take union of them. Make node 0 parent of node 1
        """
        parent = [-1] * self.total_vertices
        path = []

        # Loop over all vertices, to get all edges between src and dest
        for src_node, dest_nodes in self.adj.items():
            for dest_node in dest_nodes:
                src_parent = self.find(parent, src_node)
                dest_parent = self.find(parent, dest_node)

                path.append(dest_node)

                # If parents are the same, we have a cycle
                if src_parent == dest_parent:
                    print("Found a cycle with union-find", end=" ")
                    print(path)
                    return

                # If no cycle, combine src and dest
                self.union_vertices(parent, src_parent, dest_parent)

        return

    def directed_graph_detect_cycle_DFS(self):
        path = []
        visited = [False] * self.total_vertices
        rec_stack = [-1] * self.total_vertices

        # In an undirected graph, parent will from a loop
        # We need to keep a node's parent updated to avoid check for cycle with parent
        parent = -1

        # Call recursive helper function to detect cycle in different DFS trees
        for curr_node in range(self.total_vertices):
            if not visited[curr_node]:
                if self.directed_graph_detect_cycle_DFS_util(curr_node, visited, rec_stack, path):
                    print("Found a cycle with DFS", path)
                    break
                else:
                    print("Graph doesn't contain cycle")
        return

    def directed_graph_detect_cycle_DFS_util(self, curr_node, visited, rec_stack, path):

        if not visited[curr_node]:
            visited[curr_node] = True
            rec_stack[curr_node] = curr_node
            path.append(curr_node)

            # Recur for all adjacent vertices
            if curr_node in self.adj:
                end_node_list = self.adj[curr_node]
                for adj_node in end_node_list:
                    if not visited[adj_node]:
                        if self.undirected_graph_detect_cycle_DFS_util(adj_node, visited, path, curr_node):
                            return True
                    elif rec_stack[adj_node] >= 0:
                        path.append(adj_node)
                        print("Found a cycle with DFS", end=" ")
                        return True

        rec_stack[curr_node] = -1
        path.pop(-1)
        return False


def main():
    """
    Detect cycle in a graph
    """
    total_vertices = 7

    """
    Detect cycle in undirected graph
    Can use union-find, DFS, BFS, or graph coloring.

    Time Complexity:
        DFS: O(V + E)
        BFS: O(V + E)
        graph coloring: O(V + E)
        union-find: O(E logV)
    Space Complexity:
        All take O(V) => O(V) path of cycle, O(V) recursion stack, O(V) visited array
    """
    graph_undirected = Graph(total_vertices)
    graph_undirected.construct_undirected_graph()
    # using DFS
    graph_undirected.undirected_graph_detect_cycle_DFS()
    # using union-find
    graph_undirected.undirected_graph_detect_cycle_union_find()

    """
    Detect cycle in directed graph
    Can use DFS, tarjan's algo, or graph coloring. All take time O(V + E)

    Using DFS:
        Time Complexity: O(V + E),
        Space Complexity: O(V) => O(V) path of cycle, O(V) recursion stack, O(V) visited array
    """
    graph_directed = Graph(total_vertices)
    graph_directed.construct_directed_graph()
    # using DFS
    graph_directed.directed_graph_detect_cycle_DFS()
    return


if __name__ == '__main__':
    main()

